### PR TITLE
8268858: Determine register pressure automatically by the number of available registers for allocation

### PR DESCRIPTION
**Pull request by JoshuaZhuwj**
Hi,

This patch includes the following changes:

1) Determine register pressure automatically by the number of available
   registers for allocation.
   Currently, INTPRESSURE is hard-coded no matter whether heapbase
   or framepointer register is available for RA.

2) x86: Machines with AVX3 have 2x more XMM registers.
        The pressure threshold should be improved for both RA and LCM.
        Currently, only scheduling takes more XMM registers into
        consideration.

3) aarch64: add a new jtreg case
        In bug 8183543, this case caused many compilation failures with
        "failed spill-split-recycle sanity check". This test case is
        imported to ensure no regression failure.

Inconsistent pressure threshold value with the number of available
registers will bring unnecessary spilling in some scenarios.
I wrote a simple example as one case:
  https://cr.openjdk.java.net/~jzhu/8268858/Test.java

With this patch, 4 extra spill copies are eliminated on AArch64.
  old OptoAssembly: https://cr.openjdk.java.net/~jzhu/8268858/old_opto_assembly
  new OptoAssembly: https://cr.openjdk.java.net/~jzhu/8268858/new_opto_assembly

Could you please help review this patch?

Best Regards,
Joshua

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268858](https://bugs.openjdk.java.net/browse/JDK-8268858): Determine register pressure automatically by the number of available registers for allocation

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4505/head:pull/4505` \
`$ git checkout pull/4505`

Update a local copy of the PR: \
`$ git checkout pull/4505` \
`$ git pull https://git.openjdk.java.net/jdk pull/4505/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4505`

View PR using the GUI difftool: \
`$ git pr show -t 4505`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4505.diff">https://git.openjdk.java.net/jdk/pull/4505.diff</a>

</details>

### CODE REVIEW

**Comment by JoshuaZhuwj** [comment]
> This patch includes the following changes:
> 
> 1. Determine register pressure automatically by the number of available
>    registers for allocation.
>    Currently, INTPRESSURE is hard-coded no matter whether heapbase
>    or framepointer register is available for RA.
For those machines that I don’t have, I can’t perform architecture-dependent
tests, therefore the threshold ​​remains unchanged.

> 2. x86: Machines with AVX3 have 2x more XMM registers.
>    The pressure threshold should be improved for both RA and LCM.
>    Currently, only scheduling takes more XMM registers into
>    consideration.

> 3. aarch64: add a new jtreg case
>    In bug 8183543, this case caused many compilation failures with
>    "failed spill-split-recycle sanity check". This test case is
>    imported to ensure no regression failure.
I introduce the test in 8183543 as a jtreg case to ensure no regression failure
for my patch.

> Inconsistent pressure threshold value with the number of available
> registers will bring unnecessary spilling in some scenarios.
> I wrote a simple example as one case:
> https://cr.openjdk.java.net/~jzhu/8268858/Test.java
> 
> With this patch, 4 extra spill copies are eliminated on AArch64.
> old OptoAssembly: https://cr.openjdk.java.net/~jzhu/8268858/old_opto_assembly
> new OptoAssembly: https://cr.openjdk.java.net/~jzhu/8268858/new_opto_assembly

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (vnkozlov) [inline src/hotspot/cpu/aarch64/aarch64.ad:2559]: Explain in comment why `-1`.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86_64.ad:1774]: Explain in comment why -2.

File: test/hotspot/jtreg/compiler/c2/Test8183543.java
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/c2/Test8183543.java:1]: Place test into `compiler/regalloc`.
We are avoiding using bug Id in test name now. Use name which explains what test does.

**Review by vnkozlov — COMMENTED** [review]
Thank you. It is very nice cleanup. I have only few comments.

**Comment by JoshuaZhuwj** [comment]
> Thank you. It is very nice cleanup. I have only few comments.

@vnkozlov Thanks a lot for your review!
I have updated the comments and placed the test into proper location as your review comments.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (vnkozlov) [reply src/hotspot/cpu/aarch64/aarch64.ad:2562]: This is not explanation. Comment should explain why with -1 the issue is fixed.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86_64.ad:1776]: Why it works best? Did you tested different values or it is required for correctness? If for correctness - why?

**Review by vnkozlov — COMMENTED** [review]
Need more information.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (JoshuaZhuwj) [reply src/hotspot/cpu/x86/x86_64.ad:1776]: Not for correctness.
The value was tested and verified in SPECjvm2008 by the author of LCM's register pressure scheduling.
Therefore I continue to use the original value.
“I did in fact experiment around with different values for both float and int pressure thresholds.  The versions in changelist work best for x64 and x86.” [1]
[1] http://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2015-September/019005.html

Comment (vnkozlov) [reply src/hotspot/cpu/x86/x86_64.ad:1776]: Okay, I got it. So originally for 64-bit FLOATPRESSURE was 14 (16 -2) and that is where `-2` came from.

Comment (vnkozlov) [reply src/hotspot/cpu/x86/x86_64.ad:1776]: But for avx512 it was 28 before and it is 30 now. I don't know which number is better for performance. But your change made difference.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (vnkozlov) [inline src/hotspot/cpu/aarch64/aarch64.ad:2562]: This is not explanation. Comment should explain why with -1 the issue is fixed. (Repeating because it could be hidden in previous conversation)

File: src/hotspot/cpu/x86/x86_64.ad
Comment (JoshuaZhuwj) [reply src/hotspot/cpu/x86/x86_64.ad]: Thanks. I update accordingly.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (JoshuaZhuwj) [reply src/hotspot/cpu/aarch64/aarch64.ad]: I add more details in comments. Please let me know if I explain clearly.

**Review by vnkozlov — APPROVED** [review]
Good. I will test it.

**Comment by vnkozlov** [comment]
Testing results from tier1-3 and Xcomp are good.
You need second review.

**Comment by JoshuaZhuwj** [comment]
@vnkozlov Thanks for your review. Also thanks a lot for helping me test on different platforms.
Could I ask for second review?

**Review by dean-long — APPROVED** [review]

**Comment by JoshuaZhuwj** [comment]
Thanks @vnkozlov @dean-long

**Pull request closed without merging**
